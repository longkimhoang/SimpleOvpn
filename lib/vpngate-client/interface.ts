import {IVpnServer} from './models';

export interface FetchServersOptions {
  signal?: AbortSignal;
}

export interface VpnServerRepository {
  vpnServers: readonly IVpnServer[];
  isFetching: boolean;
  fetchServers(options?: FetchServersOptions): void;
}
