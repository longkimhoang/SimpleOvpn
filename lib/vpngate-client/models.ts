import Realm, {ObjectSchema} from 'realm';

export interface IVpnServer {
  hostName: string;
  ipAddress: string;
  score: number;
  ping: number;
  speed: number;
  country: string;
  countryCode: string;
  operator: string;
  base64EncodedOvpnConfig: string;
}

export class VpnServer extends Realm.Object implements IVpnServer {
  _id!: Realm.BSON.ObjectId;
  hostName!: string;
  ipAddress!: string;
  score!: number;
  ping!: number;
  speed!: number;
  country!: string;
  countryCode!: string;
  operator!: string;
  base64EncodedOvpnConfig!: string;

  static generate(
    properties: IVpnServer,
  ): RealmInsertionModel<VpnServer> {
    return {
      ...properties,
      _id: new Realm.BSON.ObjectId(),
    };
  }

  static schema: ObjectSchema = {
    name: 'VpnServer',
    primaryKey: '_id',
    properties: {
      _id: 'objectId',
      hostName: 'string',
      ipAddress: 'string',
      score: {
        type: 'int',
        indexed: true,
      },
      ping: 'int',
      country: 'string',
      countryCode: 'string',
      operator: 'string',
      base64EncodedOvpnConfig: 'string',
    },
  };
}
