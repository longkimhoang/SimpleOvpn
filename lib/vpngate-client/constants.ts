export const VPN_GATE_SERVERS_LIST_URL = 'https://www.vpngate.net/api/iphone/';
