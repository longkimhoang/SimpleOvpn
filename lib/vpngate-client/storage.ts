import {useMMKV, useMMKVObject} from 'react-native-mmkv';
import {IVpnServer} from './models';

/**
 * A type that manages local storage of fetched VPN servers
 */
export interface VpnServerStorage {
  data: readonly IVpnServer[];
  write(data: readonly IVpnServer[]): void | Promise<void>;
}

export function useMmkvVpnServerStorage(): VpnServerStorage {
  const storage = useMMKV({
    id: 'vpngate',
  });

  const [servers, setServers] = useMMKVObject<readonly IVpnServer[]>(
    'vpn-servers',
    storage,
  );

  return {
    data: servers ?? [],
    write: setServers,
  };
}
