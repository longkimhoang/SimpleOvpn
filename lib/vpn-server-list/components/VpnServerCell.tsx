import color from 'color';
import React, {PropsWithChildren} from 'react';
import {Platform, StyleSheet, View} from 'react-native';
import {
  TouchableHighlight,
  TouchableNativeFeedback,
} from 'react-native-gesture-handler';
import {Text} from 'react-native-paper';
import {useTheme} from '../../theming';
import {IVpnServer} from '../../vpngate-client/models';
import {VPN_SERVER_CELL_HEIGHT} from './constants';

export interface VpnServerCellProps {
  data: IVpnServer;
  onPress?: () => void;
  testID?: string;
}

type VpnServerCellWrapperProps = PropsWithChildren<
  Omit<VpnServerCellProps, 'data'>
>;

function VpnServerCellWrapper({
  onPress,
  children,
  ...rest
}: VpnServerCellWrapperProps) {
  const testID = 'vpn-server-list.cell';

  if (Platform.OS === 'android') {
    return (
      <TouchableNativeFeedback
        {...rest}
        style={[styles.container]}
        onPress={onPress}
        testID={testID}>
        {children}
      </TouchableNativeFeedback>
    );
  } else {
    return (
      <TouchableHighlight
        {...rest}
        style={[styles.container]}
        onPress={onPress}
        testID={testID}>
        {children}
      </TouchableHighlight>
    );
  }
}

function VpnServerCell({data, onPress, ...rest}: VpnServerCellProps) {
  const {hostName, ipAddress} = data;
  const theme = useTheme();

  const titleColor = theme.colors.onSurface;
  const descriptionColor = theme.colors.onSurfaceVariant;

  return (
    <VpnServerCellWrapper onPress={onPress} {...rest}>
      <View style={styles.row}>
        <View style={[styles.item, styles.content]}>
          <Text
            variant="bodyLarge"
            numberOfLines={1}
            style={[
              {
                color: titleColor,
              },
            ]}>
            {`${hostName}.opengw.net`}
          </Text>
          <Text
            variant="bodyMedium"
            numberOfLines={2}
            style={[
              {
                color: descriptionColor,
              },
            ]}>
            {ipAddress}
          </Text>
        </View>
      </View>
    </VpnServerCellWrapper>
  );
}

const styles = StyleSheet.create({
  container: {
    padding: 8,
    height: VPN_SERVER_CELL_HEIGHT,
  },
  row: {
    flexDirection: 'row',
  },
  item: {
    marginVertical: 6,
    paddingLeft: 8,
  },
  content: {
    flex: 1,
    justifyContent: 'center',
  },
});

export default React.memo(VpnServerCell);
