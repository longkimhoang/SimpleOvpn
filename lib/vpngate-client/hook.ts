import csv from 'csvtojson';
import {useCallback, useState} from 'react';
import compactMap from '../common/compactMap';
import RealmContext from '../db/realmContext';
import {VPN_GATE_SERVERS_LIST_URL} from './constants';
import {VpnServerRepository} from './interface';
import {IVpnServer, VpnServer} from './models';

//#region Typedefs

export type FetchVpnServersAction = () => Promise<readonly IVpnServer[]>;

export interface UseVpnGateClientDeps {
  fetchVpnServersAction: FetchVpnServersAction;
}

//#endregion

//#region Fetch and parse VPNGate response

const converter = csv();

function convertToVpnServer(value: any): IVpnServer {
  return {
    hostName: value['#HostName'],
    ipAddress: value['IP'],
    score: parseInt(value['Score']),
    ping: parseInt(value['Ping']),
    speed: parseInt(value['Speed']),
    country: value['CountryLong'],
    countryCode: value['CountryShort'],
    operator: value['Operator'],
    base64EncodedOvpnConfig: value['OpenVPN_ConfigData_Base64'],
  };
}

export const fetchVpnGateServers: FetchVpnServersAction = async () => {
  const response = await fetch(VPN_GATE_SERVERS_LIST_URL);
  const body = await response.text();

  const lines = body.split(/\r?\n/).map(s => s.trim());
  const data = lines.slice(1).join('\n');

  return converter
    .fromString(data)
    .then(values => compactMap(values, convertToVpnServer));
};

//#endregion

//#region Hook

export function useVpnGateClient(
  deps: UseVpnGateClientDeps = {
    fetchVpnServersAction: fetchVpnGateServers,
  },
): VpnServerRepository {
  const {fetchVpnServersAction} = deps;
  const {useRealm, useQuery} = RealmContext;

  const [isFetching, setIsFetching] = useState(() => false);
  const realm = useRealm();
  const result = useQuery(VpnServer);
  const dispatch = useCallback(() => {
    setIsFetching(true);

    fetchVpnServersAction().then(vpnServers => {
      realm.write(() => {
        for (const server of vpnServers) {
          realm.create(VpnServer, VpnServer.generate(server));
        }
      });

      setIsFetching(false);
    });
  }, []);

  return {
    vpnServers: result,
    isFetching,
    fetchServers: dispatch,
  };
}

//#endregion
