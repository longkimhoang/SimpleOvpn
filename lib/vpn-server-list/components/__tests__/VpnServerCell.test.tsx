import {render} from '@testing-library/react-native';
import React from 'react';
import {IVpnServer} from '../../../vpngate-client/models';
import VpnServerCell from '../VpnServerCell';

describe('VpnServerCell', () => {
  const mockVpnServer: IVpnServer = {
    hostName: 'public-vpn-90',
    ipAddress: '219.100.37.55',
    score: 3037585,
    ping: 16,
    speed: 71957702,
    country: 'Japan',
    countryCode: 'JP',
    operator: 'Daiyuu Nobori_ Japan. Academic Use Only.',
    base64EncodedOvpnConfig: 'mockBase64OvpnConfig',
  };

  test('should render correctly', () => {
    const tree = render(<VpnServerCell data={mockVpnServer} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('should format hostName correctly', () => {
    const {queryByText} = render(<VpnServerCell data={mockVpnServer} />);

    expect(queryByText('public-vpn-90.opengw.net')).not.toBeNull();
  });
});
