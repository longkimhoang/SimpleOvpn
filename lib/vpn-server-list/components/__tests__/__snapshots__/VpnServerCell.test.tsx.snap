// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VpnServerCell should render correctly 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": true,
    }
  }
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      false,
      Array [
        Object {
          "padding": 8,
        },
        Array [
          Object {
            "height": 60,
          },
        ],
      ],
    ]
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "marginVertical": 6,
            "paddingLeft": 8,
          },
          Object {
            "flex": 1,
            "justifyContent": "center",
          },
        ]
      }
    >
      <Text
        numberOfLines={1}
        selectable={false}
        style={
          Array [
            Object {
              "color": "rgba(28, 27, 31, 1)",
            },
            Object {
              "textAlign": "left",
            },
            Array [
              Object {
                "fontSize": 16,
              },
              Object {
                "color": "rgba(28, 27, 31, 1)",
              },
              undefined,
            ],
          ]
        }
      >
        public-vpn-90.opengw.net
      </Text>
      <Text
        numberOfLines={2}
        selectable={false}
        style={
          Array [
            Object {
              "color": "rgba(28, 27, 31, 1)",
            },
            Object {
              "textAlign": "left",
            },
            Array [
              Object {
                "fontSize": 14,
              },
              Object {
                "color": "rgba(73, 69, 79, 1)",
              },
              undefined,
            ],
          ]
        }
      >
        219.100.37.55
      </Text>
    </View>
  </View>
</View>
`;
